File: private_dot_config/nvim/lua/rc/config/ddu.lua
local M = {}

function M.setup()
  -- key bindings
  vim.keymap.set("n", "<Space>fb", "<Cmd>Ddu -name=buffer buffer<CR>", { silent = true })
  vim.keymap.set(
    "n",
    "<Space>ff",
    "<Cmd>Ddu -name=files file_point file_old file_rec -ui-param-displaySourceName=short<CR>",
    { silent = true }
  )
  vim.keymap.set("n", "/", "<Cmd>Ddu -name=search line -ui-param-startFilter<CR>")
  vim.keymap.set("n", "*", "<Cmd>Ddu -name=search line -input=`expand('<cword>')` -ui-param-startFilter=v:false<CR>")

  -- ftplugin for ddu buffers
  vim.api.nvim_create_autocmd("FileType", {
    group = "MyAutoCmd",
    pattern = "ddu-ff",
    callback = function()
      require("rc.config.ddu").ftplugin_ddu_ff()
    end,
    desc = "Set key mappings in ddu-ff buffers",
  })

  vim.api.nvim_create_autocmd("FileType", {
    group = "MyAutoCmd",
    pattern = "ddu-ff-filter",
    callback = function()
      require("rc.config.ddu").ftplugin_ddu_ff_filter()
    end,
    desc = "Set key mappings in ddu-ff-filter buffers",
  })

  vim.api.nvim_create_autocmd("FileType", {
    group = "MyAutoCmd",
    pattern = "ddu-filter",
    callback = function()
      require("rc.config.ddu").ftplugin_ddu_filer()
    end,
    desc = "Set key mappings in ddu-filer buffers",
  })
end

function M.config()
  vim.fn["ddu#custom#alias"]("source", "file_rg", "file_external")

  vim.fn["ddu#custom#patch_global"] {
    ui = "ff",
    sourceOptions = {
      ["_"] = {
        ignoreCase = true,
        matchers = { "matcher_substring" },
      },
      file_old = {
        matchers = { "matcher_substring", "matcher_relative", "matcher_hidden" },
      },
      file_external = {
        matchers = { "matcher_substring", "matcher_hidden" },
      },
      file_rec = {
        matchers = { "matcher_substring", "matcher_hidden" },
      },
    },
    sourceParams = {
      file_external = {
        cmd = { "git", "ls-files", "-co", "--exclude-standard" },
      },
    },
    uiParams = {
      ff = {
        split = "floating",
      },
    },
    kindOptions = {
      file = { defaultAction = "open" },
      word = { defaultAction = "append" },
      action = { defaultAction = "do" },
    },
  }

  -- vim.fn["ddu#custom#patch_local"]("files", {
  --   uiParams = {
  --     ff = { split = "floating" },
  --   },
  -- })

  vim.fn["ddu#custom#patch_global"] {
    sourceParams = {
      file_rg = {
        cmd = { "rg", "--files", "--glob", "!.git", "--color", "never", "--no-messages" },
        updataItems = 50000,
      },
      rg = {
        args = { "--ignore-case", "--column", "--no-heading", "--color", "never" },
      },
    },
  }

  vim.fn["ddu#custom#patch_global"] {
    filterParams = {
      matcher_substring = { highlightMatched = "Search" },
    },
  }
end

--- Settings in ddu-ff-filter buffer
function M.ftplugin_ddu_ff()
  local opts = { buffer = true }
  vim.keymap.set("n", "<CR>", "<Cmd>call ddu#ui#ff#do_action('itemAction')<CR>", opts)
  vim.keymap.set("n", "<Space>", "<Cmd>call ddu#ui#ff#do_action('toggleSelectItem')<CR>", opts)
  vim.keymap.set("n", "<Tab>", "<Cmd>call ddu#ui#ff#do_action('chooseAction')<CR>", opts)
  vim.keymap.set("n", "<C-l>", "<Cmd>call ddu#ui#ff#do_action('refreshItems')<CR>", opts)
  vim.keymap.set("n", "i", "<Cmd>call ddu#ui#ff#do_action('openFilterWindow')<CR>", opts)
  vim.keymap.set("n", "p", "<Cmd>call ddu#ui#ff#do_action('preview')<CR>", opts)
  vim.keymap.set("n", "q", "<Cmd>call ddu#ui#ff#do_action('quit')<CR>", opts)
  vim.keymap.set("n", "c", "<Cmd>call ddu#ui#ff#do_action('itemAction', {'name': 'cd'})<CR>", opts)
  vim.keymap.set("n", "d", "<Cmd>call ddu#ui#ff#do_action('itemAction', {'name': 'delete'})<CR>", opts)
  vim.keymap.set("n", "e", "<Cmd>call ddu#ui#ff#do_action('itemAction', {'name': 'edit'})<CR>", opts)
  vim.keymap.set("n", "E", "<Cmd>call ddu#ui#ff#do_action('itemAction', {'params': eval(input('params: '))})<CR>", opts)
  vim.keymap.set(
    "n",
    "v",
    "<Cmd>call ddu#ui#ff#do_action('itemAction', {'name': 'open', 'params': {'command': 'vsplit'}})<CR>",
    opts
  )
  vim.keymap.set("n", "N", "<Cmd>call ddu#ui#ff#do_action('itemAction', {'name': 'new'})<CR>", opts)
  vim.keymap.set("n", "r", "<Cmd>call ddu#ui#ff#do_action('itemAction', {'name': 'quickfix'})<CR>", opts)
  vim.keymap.set(
    "n",
    "u",
    "<Cmd>call ddu#ui#ff#do_action('updateOptions', {'sourceOptions': {'_': {'matchers': []}}})<CR>",
    opts
  )
end

--- Settings in ddu-ff-filter buffer
function M.ftplugin_ddu_ff_filter()
  local opts = { silent = true, buffer = true }
  vim.keymap.set("i", "<CR>", "<Esc><Cmd>close<CR>", opts)
  vim.keymap.set("i", "jj", "<Esc><Cmd>close<CR>", opts)
  vim.keymap.set("n", "<CR>", "<Cmd>close<CR>", opts)
  vim.keymap.set("n", "q", "<Cmd>close<CR>", opts)
end

--- Settings in ddu-filer buffer
function M.ftplugin_ddu_filer()
  vim.keymap.set("n", "q", "<Cmd>call ddu#ui#filer#do_action('quit')<CR>", { buffer = true })
  vim.keymap.set("n", "<CR>", function()
    if vim.fn["ddu#ui#filer#is_directory"]() then
      vim.fn["ddu#ui#do_action"]("itemAction", { name = "narrow" })
    else
      vim.fn["ddu#ui#do_action"]("itemAction", { name = "open" })
    end
  end, { buffer = true, expr = true })
end

return M

File: private_dot_config/nvim_dpp/dpp/ddu.ts
import {
  type ActionArguments,
  ActionFlags,
  type DduOptions,
} from "jsr:@shougo/ddu-vim@~6.1.0/types";

import {
  BaseConfig,
  type ConfigArguments,
} from "jsr:@shougo/ddu-vim@~6.1.0/config";

import { type ActionData as FileAction } from "jsr:@shougo/ddu-kind-file@~0.9.0";
import { type Params as FfParams } from "jsr:@shougo/ddu-ui-ff@~1.4.0";
import { type Params as FilerParams } from "jsr:@shougo/ddu-ui-filer@~1.4.0";

import type { Denops } from "jsr:@denops/std@~7.1.0";
import * as fn from "jsr:@denops/std@~7.1.0/function";

type Params = Record<string, unknown>;

type DppAction = {
  path: string;
  __name: string;
};

export class Config extends BaseConfig {
  override config(args: ConfigArguments): Promise<void> {
    args.setAlias("source", "file_rg", "file_external");
    args.setAlias("source", "file_git", "file_external");
    args.setAlias(
      "filter",
      "matcher_ignore_current_buffer",
      "matcher_ignores",
    );
    args.setAlias("action", "tabopen", "open");

    args.contextBuilder.patchGlobal({
      ui: "ff",
      profile: false,
      uiOptions: {
        ff: {
          actions: {
            kensaku: async (args: {
              denops: Denops;
              options: DduOptions;
            }) => {
              await args.denops.dispatcher.updateOptions(
                args.options.name,
                {
                  sourceOptions: {
                    _: {
                      matchers: ["matcher_kensaku"],
                    },
                  },
                },
              );
              await args.denops.cmd(
                "echomsg 'change to kensaku matcher'",
              );

              return ActionFlags.Persist;
            },
          },
        },
        filer: {
          toggle: true,
        },
      },
      uiParams: {
        ff: {
          autoAction: {
            name: "preview",
          },
          floatingBorder: "single",
          floatingTitle: "Results",
          floatingTitlePos: "center",
          ignoreEmpty: true,
          inputFunc: "cmdline#input",
          previewFloating: true,
          previewFloatingBorder: "single",
          previewSplit: "vertical",
          previewFloatingTitle: "Preview",
          previewFloatingTitlePos: "center",
          previewWindowOptions: [
            ["&signcolumn", "no"],
            ["&foldcolumn", 0],
            ["&foldenable", 0],
            ["&number", 0],
            ["&wrap", 0],
            ["&scrolloff", 0],
          ],
          prompt: "",
          split: "floating",
          startAutoAction: true,
        } as Partial<FfParams>,
        filer: {
          previewFloating: true,
          sort: "filename",
          sortTreesFirst: true,
          split: "no",
          toggle: true,
        } as Partial<FilerParams>,
      },
      sourceOptions: {
        _: {
          ignoreCase: true,
          matchers: ["matcher_substring"],
          smartCase: true,
        },
        buffer: {
          converters: ["converter_hl_dir"],
          columns: ["icon_filename"],
        },
        file_old: {
          matchers: [
            "matcher_relative",
            "matcher_substring",
          ],
          converters: ["converter_hl_dir"],
          columns: ["icon_filename"],
        },
        file_git: {
          matchers: [
            "matcher_relative",
            "matcher_substring",
          ],
          sorters: ["sorter_mtime"],
          converters: ["converter_hl_dir"],
          columns: ["icon_filename"],
        },
        file_rec: {
          matchers: [
            "matcher_substring",
            "matcher_hidden",
          ],
          sorters: ["sorter_mtime"],
          converters: [
            "converter_hl_dir",
          ],
          columns: ["icon_filename"],
        },
        file_fd: {
          matchers: [
            "matcher_substring",
            "matcher_hidden",
          ],
          converters: ["converter_hl_dir"],
          columns: ["icon_filename"],
        },
        file_rg: {
          matchers: [
            "matcher_substring",
            "matcher_hidden",
          ],
          converters: ["converter_hl_dir"],
          columns: ["icon_filename"],
        },
        file: {
          matchers: [
            "matcher_substring",
            "matcher_hidden",
          ],
          sorters: ["sorter_alpha"],
          converters: ["converter_hl_dir"],
          columns: ["icon_filename"],
        },
        dpp: {
          defaultAction: "cd",
          actions: {
            update: async (args: ActionArguments<Params>) => {
              const names = args.items.map((item) =>
                (item.action as DppAction).__name
              );

              await args.denops.call(
                "dpp#async_ext_action",
                "installer",
                "update",
                { names },
              );

              return Promise.resolve(ActionFlags.None);
            },
          },
        },
        command_args: {
          defaultAction: "execute",
        },
        help: {
          defaultAction: "open",
        },
        markdown: {
          sorters: [],
        },
        line: {
          matchers: [
            "matcher_kensaku",
          ],
        },
        path_history: {
          defaultAction: "uiCd",
        },
        rg: {
          matchers: [
            "matcher_substring",
            "matcher_files",
          ],
          sorters: ["sorter_alpha"],
        },
      },
      sourceParams: {
        file_git: {
          cmd: ["git", "ls-files", "-co", "--exclude-standard"],
        },
        rg: {
          args: [
            "--ignore-case",
            "--column",
            "--no-heading",
            "--color",
            "never",
          ],
        },
        file_rg: {
          cmd: [
            "rg",
            "--files",
            "--glob",
            "!.git",
            "--color",
            "never",
            "--no-messages",
          ],
          updateItems: 50000,
        },
      },
      filterParams: {
        matcher_kensaku: {
          highlightMatched: "Search",
        },
        matcher_substring: {
          highlightMatched: "Search",
        },
        matcher_ignore_files: {
          ignoreGlobs: ["test_*.vim"],
          ignorePatterns: [],
        },
        converter_hl_dir: {
          hlGroup: ["Directory", "Keyword"],
        },
      },
      kindOptions: {
        file: {
          defaultAction: "open",
          actions: {
            grep: async (args: ActionArguments<Params>) => {
              const action = args.items[0]?.action as FileAction;

              await args.denops.call("ddu#start", {
                name: args.options.name,
                push: true,
                sources: [
                  {
                    name: "rg",
                    params: {
                      path: action.path,
                      input: await fn.input(
                        args.denops,
                        "Pattern: ",
                      ),
                    },
                  },
                ],
              });

              return Promise.resolve(ActionFlags.None);
            },
          },
        },
        word: {
          defaultAction: "append",
        },
        deol: {
          defaultAction: "switch",
        },
        action: {
          defaultAction: "do",
        },
        readme_viewer: {
          defaultAction: "open",
        },
        url: {
          defaultAction: "browse",
        },
      },
      kindParams: {},
      actionOptions: {
        narrow: {
          quit: false,
        },
        tabopen: {
          quit: false,
        },
      },
    });

    return Promise.resolve();
  }
}
